runtime/src/main/java/org/corfudb/runtime/DistributedCompactor.java

```diff
@@ -0,0 +1,553 @@
+ * only opened tables are checkpointed.
+ * To checkpoint a table -
+ * 1. Check if the table is locked for checkpointing by any other client ie. check if the Status of the table is IDLE
+ * 2. Acquire lock to checkpoint the table, ie. mark table's status as STARTED if no other client already did
+ * 3. Perform checkpointing for the table's stream
+ * 4. Release lock to checkpoint the table ie. mark table's status as FAILED/COMPLETED
+ */
+
+@Slf4j
+public class DistributedCompactor {
+    private final CorfuRuntime runtime;
+    private final CorfuRuntime cpRuntime;
+    private final String persistedCacheRoot;
+    private final boolean isClient;
+
+    private final String clientName;
+
+    public static final long CONN_RETRY_DELAY_MILLISEC = 500;
+    public static final String COMPACTION_MANAGER_TABLE_NAME = "CompactionManager";
+    public static final String CHECKPOINT_STATUS_TABLE_NAME = "CheckpointStatusTable";
+    public static final String ACTIVE_CHECKPOINTS_TABLE_NAME = "ActiveCheckpoints";
+
+    public static final String CHECKPOINT = "checkpoint";
+
+    public static final StringKey COMPACTION_MANAGER_KEY = StringKey.newBuilder().setKey("CompactionManagerKey").build();
+    public static final StringKey CHECKPOINT_KEY = StringKey.newBuilder().setKey("minCheckpointToken").build();
+    public static final StringKey UPGRADE_KEY = StringKey.newBuilder().setKey("UpgradeKey").build();
+
+    private CorfuStore corfuStore = null;
+    private Table<StringKey, CheckpointingStatus, Message> compactionManagerTable = null;
+    private Table<TableName, CheckpointingStatus, Message> checkpointingStatusTable = null;
+    private Table<TableName, ActiveCPStreamMsg, Message> activeCheckpointsTable = null;
+    private Table<StringKey, TokenMsg, Message> checkpointTable = null;
+
+    private CheckpointLivenessUpdater livenessUpdater;
+
+    public static class CorfuTableNamePair {
+        public TableName tableName;
+        public CorfuTable corfuTable;
+        public CorfuTableNamePair(TableName tableName, CorfuTable corfuTable) {
+            this.tableName = tableName;
+            this.corfuTable = corfuTable;
+        }
+    }
+
+    public DistributedCompactor(CorfuRuntime corfuRuntime, CorfuRuntime cpRuntime, String persistedCacheRoot) {
+        this.runtime = corfuRuntime;
+        this.cpRuntime = cpRuntime;
+        this.persistedCacheRoot = persistedCacheRoot;
+        this.isClient = false;
+        this.clientName = corfuRuntime.getParameters().getClientName();
+    }
+
+    public DistributedCompactor(CorfuRuntime corfuRuntime) {
+        this.runtime = corfuRuntime;
+        this.cpRuntime = null;
+        this.persistedCacheRoot = null;
+        this.isClient = true;
+        this.clientName = corfuRuntime.getParameters().getClientName();
+    }
+
+    private void openCheckpointingMetadataTables() {
+        try {
+            if (this.corfuStore != null) { // only run this method once
```

Comment:
instead of this check, can we move the logic to open the metadata tables in the constructor or at init so that it is only executed once?

Reply:
we can't because some of this code, if executed in the constructor, causes a deadlock because runtime isn't instantiated in those places. That is why this check has to be made to have it done at runtime instead.